### PR TITLE
Enable a custom ReplicationTracker override per index

### PR DESCRIPTION
**Pull request by original-brownbear**
Necessary to disable writes to replicas via a plugin for stateless.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/IndexModuleTests.java [hunk 1]
```diff
@@ -718,6 +730,79 @@ public void onIndexCommitDelete(ShardId shardId, IndexCommit deletedCommit) {
         }
     }
 
+    public void testCustomRecoverySourceFactory() throws IOException {
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/IndexModuleTests.java:733]: I think you mean ReplicationTrackerFactory?

```suggestion
    public void testCustomReplicationTrackerFactory() throws IOException {
```

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/test/java/org/elasticsearch/index/IndexModuleTests.java [hunk 2]
```diff
@@ -718,6 +730,79 @@ public void onIndexCommitDelete(ShardId shardId, IndexCommit deletedCommit) {
...
+                    globalCheckpoint,
+                    onGlobalCheckpointUpdated,
+                    currentTimeMillisSupplier,
+                    onSyncRetentionLeases,
+                    safeCommitInfoSupplier,
+                    onReplicationGroupUpdated
+                );
+            }
+        }
+        module.setReplicationTrackerFactory(
```
Comment (tlrx) [inline server/src/test/java/org/elasticsearch/index/IndexModuleTests.java:771]: Can you update `testFrozen()` as well to ensure it cannot be set once the IndexModule is frozen?

**Review by tlrx — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks both!

**Pull request merged**